File: apps/web/src/lib/cta-localization.ts
/**
 * Caption / CTA language harmonization (safety net for auto-produce).
 * Mirrors backend/app/crew/cta_localization.py — keep in sync for key CTAs.
 */

const CTA_TR_TO_EN: Record<string, string> = {
  'hemen incele': 'Explore now',
  'detaylari incele': 'See details',
  'detayları incele': 'See details',
  detaylar: 'See details',
  'rezervasyon yap': 'Book now',
  'hemen rezervasyon': 'Book now',
  'masani ayir': 'Reserve a table',
  'masanı ayır': 'Reserve a table',
  'randevu al': 'Book an appointment',
  'yerini ayir': 'Save your spot',
  'yerini ayır': 'Save your spot',
  'yerini ayirt': 'Reserve your spot',
  'yerini ayırt': 'Reserve your spot',
  'hemen siparis ver': 'Order now',
  'hemen sipariş ver': 'Order now',
  'siparis ver': 'Order now',
  'sipariş ver': 'Order now',
  incele: 'Explore',
  'bizi ziyaret et': 'Visit us',
  'yolunu dusur': 'Stop by',
  'yolunu düşür': 'Stop by',
  'iletisime gec': 'Get in touch',
  'iletişime geç': 'Get in touch',
  'bilgi al': 'Get info',
  'bize katil': 'Join us',
  'bize katıl': 'Join us',
  'takip et': 'Follow us',
  'bugun dene': 'Try today',
  'bugün dene': 'Try today',
  'menuyu gor': 'View menu',
  'menüyü gör': 'View menu',
  kacirma: "Don't miss out",
  kaçırma: "Don't miss out",
  kesfet: 'Discover',
  keşfet: 'Discover',
  'daha fazla': 'Learn more',
};

const CTA_EN_TO_TR: Record<string, string> = {
  'explore now': 'Hemen incele',
  explore: 'İncele',
  'learn more': 'Detayları incele',
  'see details': 'Detayları incele',
  'discover more': 'Keşfet',
  discover: 'Keşfet',
  'book now': 'Rezervasyon yap',
  'reserve now': 'Rezervasyon yap',
  'reserve a table': 'Masanı ayır',
  'book an appointment': 'Randevu al',
  'save your spot': 'Yerini ayır',
  'reserve your spot': 'Yerini ayırt',
  'order now': 'Sipariş ver',
  'visit us': 'Bizi ziyaret et',
  'stop by': 'Yolunu düşür',
  'get in touch': 'İletişime geç',
  'contact us': 'İletişime geç',
  'get info': 'Bilgi al',
  'join us': 'Bize katıl',
  'follow us': 'Takip et',
  'try today': 'Bugün dene',
  'view menu': 'Menüyü gör',
  "don't miss out": 'Kaçırma',
  'dont miss out': 'Kaçırma',
  'check it out': 'Hemen incele',
};

const TR_FOLD: Record<string, string> = {
  İ: 'i', I: 'i', ı: 'i',
  Ö: 'o', ö: 'o',
  Ü: 'u', ü: 'u',
  Ş: 's', ş: 's',
  Ç: 'c', ç: 'c',
  Ğ: 'g', ğ: 'g',
};

function normalizeCtaKey(text: string): string {
  let s = text.trim();
  for (const [from, to] of Object.entries(TR_FOLD)) {
    s = s.split(from).join(to);
  }
  return s.toLowerCase().normalize('NFKD').replace(/\u0307/g, '');
}

function stripCtaPhrases(text: string): string {
  let cleaned = text;
  const phrases = [...Object.keys(CTA_TR_TO_EN), ...Object.keys(CTA_EN_TO_TR)];
  for (const phrase of phrases) {
    cleaned = cleaned.replace(new RegExp(phrase.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'gi'), ' ');
  }
  return cleaned;
}

export function detectCtaLanguage(text: string): 'en' | 'tr' {
  const raw = text.trim();
  if (!raw) return 'en';

  const normKey = normalizeCtaKey(raw);
  if (normKey in CTA_TR_TO_EN) return 'tr';
  if (normKey in CTA_EN_TO_TR) return 'en';

  const sample = stripCtaPhrases(raw);
  const lower = ` ${sample.toLowerCase()} `;
  const enMarkers = [' the ', ' and ', ' your ', ' our ', ' discover ', ' why ', ' for ', ' with ', ' real ', ' about ', ' guests '];
  const enHits = enMarkers.filter((w) => lower.includes(w)).length;
  const trMarkers = [' bir ', ' ile ', ' için ', ' ve ', ' bu ', ' şimdi ', ' hemen '];
  const trHits = trMarkers.filter((w) => lower.includes(w)).length;

  if (enHits >= 2 && trHits <= 1) return 'en';
  if (trHits >= 2 && enHits <= 1) return 'tr';
  if (/[çğıöşüÇĞİÖŞÜ]/.test(sample) && enHits === 0) return 'tr';
  if (enHits >= 1) return 'en';
  return 'tr';
}

export function localizeCta(cta: string, targetLang: 'en' | 'tr'): string {
  const trimmed = cta.trim();
  if (!trimmed) return trimmed;
  const key = normalizeCtaKey(trimmed);
  if (targetLang === 'en') {
    if (detectCtaLanguage(trimmed) === 'en') return trimmed;
    return CTA_TR_TO_EN[key] ?? trimmed;
  }
  if (detectCtaLanguage(trimmed) === 'tr') return trimmed;
  return CTA_EN_TO_TR[key] ?? trimmed;
}

export function localizeCtas(ctas: string[], targetLang: 'en' | 'tr'): string[] {
  const seen = new Set<string>();
  const out: string[] = [];
  for (const c of ctas) {
    const loc = localizeCta(String(c), targetLang);
    if (!loc || seen.has(loc.toLowerCase())) continue;
    seen.add(loc.toLowerCase());
    out.push(loc);
  }
  return out;
}

export function parseCtaList(raw: unknown): string[] {
  if (raw == null) return [];
  if (Array.isArray(raw)) {
    return raw.map((c) => String(c).trim()).filter(Boolean);
  }
  if (typeof raw === 'string') {
    const text = raw.trim();
    if (!text) return [];
    if (text.startsWith('[')) {
      try {
        const parsed = JSON.parse(text) as unknown;
        if (Array.isArray(parsed)) {
          return parsed.map((c) => String(c).trim()).filter(Boolean);
        }
      } catch {
        /* fall through */
      }
    }
    if (text.includes('|')) {
      return text.split('|').map((p) => p.trim()).filter(Boolean);
    }
    return [text];
  }
  return [];
}

/**
 * Pick a CTA from brand `default_ctas` in the brand content language.
 * Prefers native-language entries, else localizes known Turkish presets.
 */
export function pickLocalizedCta(
  ctas: unknown,
  brandLanguages?: unknown,
  fallback?: string,
): string {
  const target = resolveBrandLanguageCode(brandLanguages);
  const items = parseCtaList(ctas);
  if (items.length === 0) {
    if (fallback?.trim()) return localizeCta(fallback, target);
    return target === 'en' ? 'Learn more' : 'Daha fazla';
  }
  for (const c of items) {
    if (detectCtaLanguage(c) === target) return c;
  }
  const localized = localizeCtas(items, target);
  if (localized[0]) return localized[0];
  return localizeCta(items[0]!, target);
}

function replaceEmbeddedCta(text: string, oldCta: string, newCta: string): string {
  if (!text || !oldCta || !newCta || normalizeCtaKey(oldCta) === normalizeCtaKey(newCta)) return text;

  const direct = new RegExp(oldCta.trim().replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'i');
  if (direct.test(text)) return text.replace(direct, newCta.trim());

  const oldWords = oldCta.split(/\s+/);
  const textWords = text.split(/\s+/);
  const oldNorm = oldWords.map(normalizeCtaKey);
  for (let i = 0; i <= textWords.length - oldWords.length; i++) {
    const window = textWords.slice(i, i + oldWords.length);
    if (window.map(normalizeCtaKey).join(' ') === oldNorm.join(' ')) {
      textWords.splice(i, oldWords.length, newCta.trim());
      return textWords.join(' ');
    }
  }
  return text;
}

/** Normalize brand languages from string, comma-list, or string[] contract shape. */
export function normalizeBrandLanguagesInput(raw: unknown): string {
  if (raw == null) return 'tr';
  if (Array.isArray(raw)) {
    const first = raw.map((v) => String(v).trim()).find(Boolean);
    return first ?? 'tr';
  }
  if (typeof raw === 'string') {
    const trimmed = raw.trim();
    return trimmed || 'tr';
  }
  const coerced = String(raw).trim();
  return coerced || 'tr';
}

export function resolveBrandLanguageCode(raw: unknown): 'tr' | 'en' {
  const primary = normalizeBrandLanguagesInput(raw).split(',')[0]?.trim().toLowerCase() || 'tr';
  // Match Python resolve_language_code: "en", "en-US", "english", "eng"
  if (primary === 'en' || primary.startsWith('en-') || primary.startsWith('eng')) return 'en';
  return 'tr';
}

/** Align caption + CTA to brand language. Works even when caption is empty. */
export function harmonizeCaptionAndCta(
  caption: string,
  cta: string,
  brandLanguages?: unknown,
): { caption: string; cta: string } {
  const cap = caption.trim();
  const rawCta = cta.trim();
  if (!rawCta) return { caption: cap, cta: rawCta };

  const brandLang = resolveBrandLanguageCode(brandLanguages);
  const newCta = localizeCta(rawCta, brandLang);
  if (!newCta || normalizeCtaKey(newCta) === normalizeCtaKey(rawCta)) {
    return { caption: cap, cta: rawCta };
  }
  if (!cap) return { caption: cap, cta: newCta };
  return {
    caption: replaceEmbeddedCta(cap, rawCta, newCta),
    cta: newCta,
  };
}

File: backend/app/crew/cta_localization.py
"""
CTA + caption language harmonization.

Brand default_ctas are often seeded in Turkish while content may be generated in
English (or vice versa). This module keeps caption, cta, and canva CTA fields
aligned to the brand's configured output language.
"""

from __future__ import annotations

import re
import unicodedata
from typing import Any

_TR_FOLD = str.maketrans({
    "İ": "i", "I": "i", "ı": "i",
    "Ö": "o", "ö": "o",
    "Ü": "u", "ü": "u",
    "Ş": "s", "ş": "s",
    "Ç": "c", "ç": "c",
    "Ğ": "g", "ğ": "g",
})


def _normalize_cta_key(text: str) -> str:
    """ASCII-ish fold for CTA dictionary lookup (handles Turkish İ/ı)."""
    folded = text.strip().translate(_TR_FOLD).lower()
    decomposed = unicodedata.normalize("NFKD", folded)
    return "".join(c for c in decomposed if not unicodedata.combining(c))

_LANG_MAP = {
    "en": "English",
    "tr": "Turkish",
    "de": "German",
    "fr": "French",
    "es": "Spanish",
}

# Canonical TR → EN (extend as needed)
_CTA_TR_TO_EN: dict[str, str] = {
    "hemen incele": "Explore now",
    "detayları incele": "See details",
    "detaylari incele": "See details",
    "rezervasyon yap": "Book now",
    "yerini ayırt": "Reserve your spot",
    "yerini ayirt": "Reserve your spot",
    "sipariş ver": "Order now",
    "siparis ver": "Order now",
    "iletişime geç": "Get in touch",
    "iletisime gec": "Get in touch",
    "bize katıl": "Join us",
    "bize katil": "Join us",
    "takip et": "Follow us",
    "bugün dene": "Try today",
    "bugun dene": "Try today",
    "menüyü gör": "View menu",
    "menuyu gor": "View menu",
    "kaçırma": "Don't miss out",
    "kacirma": "Don't miss out",
    "keşfet": "Discover",
    "kesfet": "Discover",
    "hemen rezervasyon": "Book now",
}

_CTA_EN_TO_TR: dict[str, str] = {
    "explore now": "Hemen incele",
    "learn more": "Detayları incele",
    "see details": "Detayları incele",
    "discover more": "Keşfet",
    "discover": "Keşfet",
    "book now": "Rezervasyon yap",
    "reserve now": "Rezervasyon yap",
    "reserve your spot": "Yerini ayırt",
    "order now": "Sipariş ver",
    "get in touch": "İletişime geç",
    "contact us": "İletişime geç",
    "join us": "Bize katıl",
    "follow us": "Takip et",
    "try today": "Bugün dene",
    "view menu": "Menüyü gör",
    "don't miss out": "Kaçırma",
    "dont miss out": "Kaçırma",
    "check it out": "Hemen incele",
}

_TURKISH_CHARS = set("çğıöşüÇĞİÖŞÜ")


def resolve_language_code(languages: str | None) -> str:
    raw = (languages or "tr").split(",")[0].strip().lower()
    if raw in _LANG_MAP:
        return raw
    if raw.startswith("eng"):
        return "en"
    if raw.startswith("tur") or raw == "türkçe":
        return "tr"
    return "en"


def resolve_output_language(languages: str | None) -> str:
    return _LANG_MAP.get(resolve_language_code(languages), "English")


def _strip_cta_phrases(text: str) -> str:
    """Remove known CTA phrases so they do not skew caption language detection."""
    cleaned = text
    for phrase in list(_CTA_TR_TO_EN.keys()) + list(_CTA_EN_TO_TR.keys()):
        pattern = re.compile(re.escape(phrase), re.IGNORECASE)
        cleaned = pattern.sub(" ", cleaned)
    return cleaned


def detect_text_language(text: str) -> str:
    """Rough heuristic: prefer word markers over Turkish chars in brand names."""
    if not text or not text.strip():
        return "en"
    raw = text.strip()
    norm_key = _normalize_cta_key(raw)
    if norm_key in _CTA_TR_TO_EN:
        return "tr"
    if norm_key in _CTA_EN_TO_TR:
        return "en"

    sample = _strip_cta_phrases(raw)
    lower = f" {sample.lower()} "

    en_markers = (
        " the ", " and ", " your ", " our ", " discover ", " why ", " for ",
        " with ", " real ", " about ", " guests ", " unforgettable ", " experiences ",
        " stories ", " smiles ", " this ", " that ", " you ", " we ",
    )
    en_hits = sum(1 for w in en_markers if w in lower)

    tr_markers = (" bir ", " ile ", " için ", " ve ", " bu ", " şimdi ", " hemen ", " kaçırma ")
    tr_hits = sum(1 for w in tr_markers if w in lower)
    tr_hits += sum(1 for w in _CTA_TR_TO_EN if w in _normalize_cta_key(sample))

    if en_hits >= 2 and tr_hits <= 1:
        return "en"
    if tr_hits >= 2 and en_hits <= 1:
        return "tr"

    if any(c in _TURKISH_CHARS for c in sample):
        tr_words = [
            w for w in re.findall(r"\b\w+\b", sample)
            if any(c in _TURKISH_CHARS for c in w)
        ]
        # Ignore capitalized proper nouns (e.g. Sarnıç Beach) when English markers present
        substantive_tr = [
            w for w in tr_words
            if w.lower() in _CTA_TR_TO_EN or (not w[0].isupper() and len(w) > 3)
        ]
        if substantive_tr or tr_hits >= 1:
            return "tr"
        if en_hits >= 1:
            return "en"
        return "tr"

    if en_hits >= 1:
        return "en"
    if tr_hits >= 1:
        return "tr"
    return "tr"


def localize_cta(cta: str, target_lang: str) -> str:
    cta = (cta or "").strip()
    if not cta:
        return cta
    key = _normalize_cta_key(cta)
    if target_lang == "en":
        if detect_text_language(cta) == "en":
            return cta
        return _CTA_TR_TO_EN.get(key, cta)
    if target_lang == "tr":
        if detect_text_language(cta) == "tr":
            return cta
        return _CTA_EN_TO_TR.get(key, cta)
    return cta


def localize_ctas(ctas: list[str], target_lang: str) -> list[str]:
    seen: set[str] = set()
    out: list[str] = []
    for c in ctas:
        loc = localize_cta(str(c), target_lang)
        if loc and loc.lower() not in seen:
            seen.add(loc.lower())
            out.append(loc)
    return out


def _replace_embedded_cta(text: str, old_cta: str, new_cta: str) -> str:
    if not text or not old_cta or not new_cta or _normalize_cta_key(old_cta) == _normalize_cta_key(new_cta):
        return text
    pattern = re.compile(re.escape(old_cta.strip()), re.IGNORECASE)
    if pattern.search(text):
        return pattern.sub(new_cta.strip(), text, count=1)
    old_words = old_cta.split()
    text_words = text.split()
    if not old_words:
        return text
    old_norm = [_normalize_cta_key(w) for w in old_words]
    for i in range(len(text_words) - len(old_words) + 1):
        window = text_words[i : i + len(old_words)]
        if [_normalize_cta_key(w) for w in window] == old_norm:
            text_words[i : i + len(old_words)] = [new_cta.strip()]
            return " ".join(text_words)
    return text


def harmonize_concept_copy(concept: dict[str, Any], languages: str | None) -> dict[str, Any]:
    """Align cta + caption fields to brand language; fix mixed-language captions."""
    target = resolve_language_code(languages)
    caption = str(concept.get("caption_draft") or concept.get("caption") or "")
    alt = str(concept.get("caption_draft_alt") or "")
    raw_cta = str(concept.get("cta") or concept.get("call_to_action") or "")

    # Brand language setting wins — never let detected caption language override tenant choice
    effective_lang = target

    old_cta = raw_cta
    new_cta = localize_cta(raw_cta, effective_lang) if raw_cta else raw_cta

    if new_cta:
        concept["cta"] = new_cta
        if old_cta and old_cta != new_cta:
            concept["caption_draft"] = _replace_embedded_cta(caption, old_cta, new_cta)
            concept["caption_draft_alt"] = _replace_embedded_cta(alt, old_cta, new_cta)
        elif caption and new_cta.lower() not in caption.lower():
            # Append CTA naturally if missing from caption body
            sep = " — " if effective_lang == "tr" else ". "
            if not caption.rstrip().endswith((".", "!", "?", "…")):
                concept["caption_draft"] = f"{caption.rstrip()}{sep}{new_cta}"
            else:
                concept["caption_draft"] = f"{caption.rstrip()} {new_cta}"

    canva = concept.get("canva_field_copy")
    if isinstance(canva, dict) and new_cta:
        canva = dict(canva)
        if canva.get("cta"):
            canva["cta"] = localize_cta(str(canva["cta"]), effective_lang)
        concept["canva_field_copy"] = canva

    return concept


def harmonize_content_concepts(
    concepts: list[dict[str, Any]],
    languages: str | None,
) -> list[dict[str, Any]]:
    return [harmonize_concept_copy(dict(c), languages) for c in concepts]
